### PR TITLE
[math] Add new Python tutorial for minimization

### PR DESCRIPTION
**Pull request by lmoneta**
Convert `NumericalMinimization.C` tutorial to Python

### CODE REVIEW

**Comment by lmoneta** [comment]
Starting build on ROOT-ubuntu2004/default with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2004/default with flags -Dtmva-sofie=On -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2004/default with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Review by couet — APPROVED** [review]
LGTM, There is no plot to show?

File: tutorials/fit/NumericalMinimization.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#  \file
+#  \ingroup tutorial_fit
+#  \notebook -nodraw
+#  Example on how to use the new Minimizer class in ROOT
+#   Show usage with all the possible minimizers.
+#  Minimize the Rosenbrock function (a 2D -function)
+#  This example is described also in
+#  http://root.cern.ch/drupal/content/numerical-minimization#multidim_minim
```
Comment (couet) [inline tutorials/fit/NumericalMinimization.py:8]: Do you really want to refer to the old drupal wb site?

**Comment by lmoneta** [comment]
No there is no plot to show, this just performing a minimization of a function and it is a translation of the same existing tutorial in C++.  We could maybe make a nice plot showing the steps to converge to the minimum

File: tutorials/fit/NumericalMinimization.py (hunk 1 above)
Comment (lmoneta) [reply tutorials/fit/NumericalMinimization.py:8]: Good point! This link is broken, we need to fix it also in the C++ version of the tutorial.

File: tutorials/fit/NumericalMinimization.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#  input : minimizer name + algorithm name
+#  randomSeed: = <0 : fixed value: 0 random with seed 0; >0 random with given seed
+#
+#  \macro_code
+#
+#  \author Lorenzo Moneta
+
+import ROOT
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:16]: This is a comment valid for the whole file. It would be better to stick to PEP8 as much as possible when writing Python code. For example, naming of variables and functions should follow snake_case and not CamelCase (apart from the calls to ROOT API which uses CamelCase).

File: tutorials/fit/NumericalMinimization.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+#     GSLMultiMin                ConjugateFR, ConjugatePR, BFGS, BFGS2, SteepestDescent
+#     GSLSimAn
+#     Genetic
+def NumericalMinimization(minimizerName = "Minuit2",
+                          algoName = "" ,
+                          randomSeed = -1) :
+
+   minimizer = ROOT.Math.Factory.CreateMinimizer(minimizerName, algoName);
+   if (not minimizer) :
+      print("Error: cannot create minimizer \"",minName,"\". Maybe the required library was not built?")
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:37]: There are a few problems here. The first is that the `minName` variable is not defined. I guess you mean `minimizerName`. Second, I don't think this way of using print is what we want:
Python 2 will print it as a tuple
```python
>>> a = "here"
>>> print("This is it: ", a, " --> hello")
('This is it: ', 'here', ' --> hello')
```
Python 3 will add unnecessary whitespace to the string.
```
>>> a = "here"
>>> print("This is it: ", a, " --> hello")
This is it:  here  --> hello
```

The more correct way would be using `format`, which is Python2 and 3 compatible:
```python
print("Error: cannot create minimizer \"{}\". Maybe the required library was not built?".format(minimizerName))
```

File: tutorials/fit/NumericalMinimization.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      return 1;
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:38]: In general returning a value that won't be used is not a common practice in Python. Especially with the mean of "error code" as used here. There is one notable exception that actually makes this practice quite useful, when used in pair with `SystemExit`. See comment on the last line for more info.

File: tutorials/fit/NumericalMinimization.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+   if ( ret and minimizer.MinValue()  < 1.E-4  and f(xs) < 1.E-4) :
+      print("Minimizer " , minimizerName," - ",algoName,
+             "converged to the right minimum!")
+   else :
+      print("Minimizer " ,minimizerName, " - ",algoName, "failed to converge !!!");
+      ROOT.Error("NumericalMinimization","fail to converge!");
+
+   return 0;
+
+NumericalMinimization()
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:85]: Calling the function uncoditionally at script level is not good practice. It would be better to use the syntax

```python
if __name__ == "__main__":
    raise SystemExit(NumericalMinimization())
```
The use of `SystemExit` makes the return values of the `NumericalMinimization` function actually useful. An alternative would be using the simpler notation:
```python
if __name__ == "__main__":
    NumericalMinimization()
```
But in this case, there should be no return values from the function. Up to you what you choose

File: tutorials/fit/NumericalMinimization.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                          algoName = "" ,
+                          randomSeed = -1) :
+
+   minimizer = ROOT.Math.Factory.CreateMinimizer(minimizerName, algoName);
+   if (not minimizer) :
+      print("Error: cannot create minimizer \"",minName,"\". Maybe the required library was not built?")
+      return 1;
+
+   # set tolerance and other minimizer parameters
+   # one can also use default values
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:41]: All comments should start with capital letter

File: tutorials/fit/NumericalMinimization.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+   minimizer.SetMaxFunctionCalls(1000000); # working for Minuit/Minuit2
+   minimizer.SetMaxIterations(10000);  # for GSL minimizers - no effect in Minuit/Minuit2
+   minimizer.SetTolerance(0.001);
+   minimizer.SetPrintLevel(1);
+
+   # create function wrapper for minimizer
+
+   f = ROOT.Math.Functor(RosenBrock,2);
+
+   #evaluate function at a point
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:51]: Comment should start with capital letter. Also, there should be a whitespace after the `#` sign

File: tutorials/fit/NumericalMinimization.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+   import numpy as np
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:52]: `import` statements should go at the top of the module

File: tutorials/fit/NumericalMinimization.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+   # Set the free variables to be minimized !
+   minimizer.SetVariable(0,"x",variable[0], step[0]);
+   minimizer.SetVariable(1,"y",variable[1], step[1]);
+
+   # do the minimization
+   ret = minimizer.Minimize();
+
+   xs = minimizer.X();
+   print("Minimum: f(", xs[0], ",", xs[1],"): ",minimizer.MinValue())
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:74]: In all print statements please use `format`, see comment above

File: tutorials/fit/NumericalMinimization.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+   if ( ret and minimizer.MinValue()  < 1.E-4  and f(xs) < 1.E-4) :
+      print("Minimizer " , minimizerName," - ",algoName,
+             "converged to the right minimum!")
+   else :
+      print("Minimizer " ,minimizerName, " - ",algoName, "failed to converge !!!");
+      ROOT.Error("NumericalMinimization","fail to converge!");
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:81]: Why do you use `ROOT.Error` here? In the error case above (the one `if (not minimizer) :`) you were just returning an error code. The tutorial should be more consistent regarding error handling.

If you decide to keep the error here, it should be a `raise RuntimeError` or other standard Python exceptions.

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thanks for adding the tutorial! See some comments before merging

**Comment by lmoneta** [comment]
Thank you very much @vepadulano for all the comments. I have included them in the new version.

File: tutorials/fit/NumericalMinimization.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+#  randomSeed: = <0 : fixed value: 0 random with seed 0; >0 random with given seed
+#
+#  \macro_code
+#
+#  \author Lorenzo Moneta
+
+import ROOT
+import numpy as np
+
+def RosenBrock( vecx ):
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:18]: The file is still missing a pass with a formatter such as `autopep8`

File: tutorials/fit/NumericalMinimization.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+# create minimizer giving a name and a name (optionally) for the specific algorithm
+#  possible choices are:
+#     minimizerName                  algoName
+#
+#     Minuit                     Migrad, Simplex,Combined,Scan  (default is Migrad)
+#     Minuit2                    Migrad, BFGS, Simplex,Combined,Scan  (default is Migrad)
+#     GSLMultiMin                ConjugateFR, ConjugatePR, BFGS, BFGS2, SteepestDescent
+#     GSLSimAn
+#     Genetic
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:31]: These comments should become a docstring, right?

File: tutorials/fit/NumericalMinimization.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+   # Set the free variables to be minimized !
+   minimizer.SetVariable(0,"x",variable[0], step[0]);
+   minimizer.SetVariable(1,"y",variable[1], step[1]);
+
+   # Do the minimization
+   ret = minimizer.Minimize();
+
+   xs = minimizer.X();
+   print("Minimum: f(", xs[0], ",", xs[1],"): ",minimizer.MinValue())
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:74]: This print statement is missing `format`

File: tutorials/fit/NumericalMinimization.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+   # Real minimum is f(xmin) = 0
+   if ( ret and minimizer.MinValue()  < 1.E-4 ) :
+      print("Minimizer {} - {} converged to the right minimum!".format(minimizerName,algoName))
```
Comment (vepadulano) [inline tutorials/fit/NumericalMinimization.py:78]: I suppose this is the "correct" output from the function right? Is there no correct return value? Does the tutorial only show this print statement? What is the user supposed to see at the end of the tutorial?

**Review by vepadulano — CHANGES_REQUESTED** [review]

File: tutorials/fit/NumericalMinimization.py (hunk 13 above)
Comment (lmoneta) [reply tutorials/fit/NumericalMinimization.py]: Thanks, this is also fixed now!

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-performance-centos8-multicore/cxx17 -DCTEST_TEST_EXCLUDE_NONE=On

**Review by vepadulano — APPROVED** [review]
